Optionally use neovim's api-highlights

andreypopp opened the pull request:
Fixes #1274

Took a stub at it and it seems to work well. Highlights are indeed being
adapted to line instertions so we see less stalled highlights.

I'm going to use it for a while to see how it works.

<!--
Before creating a pull request, do the following.

* Read the Contributing guide linked above first.
* Read the documentation that comes with ALE with `:help ale-development`.

Have fun!
-->

Where are the tests? Have you added tests? Have you updated the tests? Read the
comment above and the documentation referenced in it first. Write tests!

Seriously, read `:help ale-development` and write tests.

andreypopp commented:
This is how it looks:
![vim-ale-nvim](https://user-images.githubusercontent.com/30594/50538044-e86e4580-0b79-11e9-9fe9-8e5b80408ee3.gif)

RyanSquared commented:
Please check linter errors. Unfortunately, I'm not sure what to do about the errors with the outdated neovim. There's not much we can do about that unless the linter gets updated, I think?

w0rp commented on autoload/ale/highlight.vim:
```diff
@@ -49,6 +58,18 @@ endfunction
+
+else
+
 function! ale#highlight#RemoveHighlights() abort
```
I think you should move the version check inside the function here, and define it once. That will be easier to test, and won't cost too much. It looks like the changes here break earlier versions of NeoVim. You might have to check if `nvim_buf_clear_namespace` exists too. it might be that `nvim_buf_add_highlight` exists in some versions where `nvim_buf_clear_namespace` does not exist.

w0rp commented on autoload/ale/highlight.vim:
```diff
@@ -57,6 +78,79 @@ function! ale#highlight#RemoveHighlights() abort
+            let l:group = 'ALEStyleError'
+        else
+            let l:group = 'ALEError'
+        endif
```
Large chunks of code here are identical. Instead of optionally defining one function or the other, separate the code logically into the common parts between the general Vim API and the NeoVim API, and use a check at the top to call one of the two different implementations. You could, for example, build a `List` containing `Dictionary` objects with all of the data to be set, like the `l:group` value and so on, and then use that data for both implementations.

andreypopp commented:
I've been using this functionality with no problems for a couple of weeks, proceeding with it.

Thanks for the feedback @w0rp, I believe I've addressed all points.

w0rp commented on autoload/ale/highlight.vim:
```diff
@@ -26,6 +26,24 @@ endif
+let s:nvim_api = exists('*nvim_buf_add_highlight') && exists('*nvim_buf_clear_namespace')
+
+function! s:ale_nvim_highlight_id(bufnr) abort
+    if get(b:, 'ale_nvim_highlight_id', -1) == -1
```
Use `getbufvar` instead of `b:` variables. The variables for `b:` can differ from from buffer for the given buffer number.

w0rp commented on autoload/ale/highlight.vim:
```diff
@@ -26,6 +26,24 @@ endif
+function! s:ale_nvim_highlight_id(bufnr) abort
+    if get(b:, 'ale_nvim_highlight_id', -1) == -1
+        " NOTE: This will highlights nothing but will allocate new id
+        let b:ale_nvim_highlight_id = nvim_buf_add_highlight(a:bufnr, 0, '', 0, 0, -1)
```
You can use `setbufvar` here.

w0rp commented on autoload/ale/highlight.vim:
```diff
@@ -49,12 +67,83 @@ endfunction
+            call nvim_buf_add_highlight(
+                \ a:bufnr, l:highlight_id, a:group,
+                \ l:end_lnum, 0, l:end_col
+                \ )
```
Put line continuation characters on the same level of indentation as previous lines, and up-indent the content within. See most of the code in the codebase for examples.

andreypopp replied:
Fixed. We'll need a lint rule for that!

andreypopp commented:
> Is there a way to get highlights via this NeoVim API, as well as set them? If so, add tests which check that the NeoVim-specific highlights are set.

Unfortunately no at this time, this should be possible soon though (asked on neovim gitter channel).

andreypopp commented:
Lint errors addressed.

w0rp commented:
I think we should wrap all of the NeoVim API calls in functions we can replace in tests, and use those calls to capture what input is being given. That way we'll be able to test that this code works, and more importantly prevent regressions later. See most of the LSP tests for examples of mocking function calls.

w0rp commented:
Have a look at some existing tests. You have to make sure the file is loaded first with `runtime`, replace it, and then load it again in an After block.

andreypopp commented:
@w0rp is there a way to execute tests "locally" w/o docker and only a subset of tests? I find it very hard to iterate (I'm vimscript beginner!) when tests take so much time to run. I really want to finish my PRs to ALE and bring more.

w0rp commented:
If you install Vader, you can run tests for a Vader file with the `:Vader` command. You can run subset of tests with the Docker image with a glob like `./run-tests test/directory/*.vader`. Use `--help` or check the documentation to see how you can limit the Vim versions to run some tests more quickly. Make sure to install Docker for running all of the tests locally.

andreypopp commented:
This is ready for review.

AppVeyor task should be cancelled as I froze it with leaving `:echo` in test code in past commit.
For some reason tests fail locally on my neovim 0.4.0 due to not being able to override functions (I suspect neovim 0.4.0 has problems with Vader) but it's ok on CI.

davidsierradz commented:
Hi, this commit somehow needs `let g:ale_set_signs = 1` to work correct, let me explain:

Without commit `114198e`:

With `let g:ale_set_signs = 0` (good):

![2019-05-05-100823_1920x1080_scrot](https://user-images.githubusercontent.com/9190258/57195998-109df480-6f1e-11e9-8b5b-745300531223.png)

With `let g:ale_set_signs = 1` (good):

![2019-05-05-100835_1920x1080_scrot](https://user-images.githubusercontent.com/9190258/57195997-109df480-6f1e-11e9-8999-0d0d6532a716.png)

With commit `114198e`:

With `let g:ale_set_signs = 0` (bad, send error highlight to next line):

![2019-05-05-100717_1920x1080_scrot](https://user-images.githubusercontent.com/9190258/57195999-109df480-6f1e-11e9-8d98-9f8a321f7d10.png)

With `let g:ale_set_signs = 1` (good):

![2019-05-05-100637_1920x1080_scrot](https://user-images.githubusercontent.com/9190258/57196000-11368b00-6f1e-11e9-8f7f-c88fa8cba2af.png)

andreypopp commented:
@davidsierradz indeed, thanks for catching that, it's off by one error. Should be fixed with #2475
